### ISSUE TITLE
Duplicate history entries

### ISSUE THREAD

**Issue by oceans112** [qutebrowser/qutebrowser#4310]
Version info: https://paste.the-compiler.org/view/ce9eb87d

As the title goes, history page shows duplicate consecutive entries. 
It seems to happen most of the time but not always. Clearing the cache and deleting the configuration doesn't help. 
I can see those entries in the history.sqlite database, History table too, so I guess it's not a reading/rendering issue.

**Comment by The-Compiler** [comment]
That's intended - if you visited a, b, then a again, the history page should show that.

**Comment by The-Compiler** [comment]
Do you see this on pages other than GitHub? I think this is caused by [QTBUG-65223](https://bugreports.qt.io/browse/QTBUG-65223). I guess as a workaround, we could check whether the newest entry is already the same we're inserting, but I'm not sure how that'd affect performance.

cc @rcorre

**Comment by oceans112** [comment]
Yes, it happens with pretty much every site I visit. In only very few occasions the entry is single.
If you need more data, maybe a can send you the history file (22MB) or maybe I can attach some screenshot to qute:history.

**Comment by The-Compiler** [comment]
Oh, you're right! Sorry, should've tried this myself earlier... I can reproduce this (something like 1 out of 4 times) when launching with `--temp-basedir` on the default DuckDuckGo page.

Looks like `loadFinished` is emitted twice, maybe since Qt 5.11.2?

Loading:

```
11:07:21 DEBUG    webview    browsertab:_set_load_status:814 load status for <qutebrowser.browser.webengine.webenginetab.WebEngineTab tab_id=0 url='https://start.duckduckgo.com/'>: LoadStatus.loading
11:07:21 DEBUG    signals    signalfilter:_filter_signals:86 emitting: cur_load_status_changed('loading') (tab 0)
11:07:21 DEBUG    signals    signalfilter:_filter_signals:86 emitting: cur_load_started() (tab 0)
```

Finished 1:

```
[...]
11:07:22 DEBUG    sql        sql:run:207 Running SQL query: "INSERT INTO History (url, title, atime, redirect) values(:url, :title, :atime, :redirect)"
11:07:22 DEBUG    sql        sql:run:210 query bindings: {':atime': 1538903242, ':redirect': False, ':title': 'DuckDuckGo — Privacy, simplified.', ':url': 'https://start.duckduckgo.com/'}
[...]
11:07:22 DEBUG    webview    browsertab:_set_load_status:814 load status for <qutebrowser.browser.webengine.webenginetab.WebEngineTab tab_id=0 url='https://start.duckduckgo.com/'>: LoadStatus.success_https
11:07:22 DEBUG    signals    signalfilter:_filter_signals:86 emitting: cur_load_status_changed('success_https') (tab 0)
11:07:22 DEBUG    signals    signalfilter:_filter_signals:86 emitting: cur_load_finished(True) (tab 0)
```

Finished 2:

```
[...]
11:07:22 DEBUG    sql        sql:run:207 Running SQL query: "INSERT INTO History (url, title, atime, redirect) values(:url, :title, :atime, :redirect)"
11:07:22 DEBUG    sql        sql:run:210 query bindings: {':atime': 1538903242, ':redirect': False, ':title': 'DuckDuckGo — Privacy, simplified.', ':url': 'https://start.duckduckgo.com/'}
[...]
11:07:22 DEBUG    webview    browsertab:_set_load_status:814 load status for <qutebrowser.browser.webengine.webenginetab.WebEngineTab tab_id=0 url='https://start.duckduckgo.com/'>: LoadStatus.success_https
11:07:22 DEBUG    signals    signalfilter:_filter_signals:86 emitting: cur_load_status_changed('success_https') (tab 0)
11:07:22 DEBUG    signals    signalfilter:_filter_signals:86 emitting: cur_load_finished(True) (tab 0)
```

Wonder if https://codereview.qt-project.org/#/c/239214/ changes anything, though that's for 5.12. Still might be our workaround for that Qt bug, though...

**Comment by rcorre** [comment]
> With Qt 5.10, everything seems fine...

Dunno if this is relevant, but with Qt 5.10 we apparently 'fake' `load_finished`:

```py
        # WORKAROUND for https://bugreports.qt.io/browse/QTBUG-65223
        if qtutils.version_check('5.10', compiled=False):
            page.loadProgress.connect(self._on_load_progress_workaround)
            self._load_finished_fake.connect(self._on_history_trigger)
            self._load_finished_fake.connect(self._restore_zoom)
            self._load_finished_fake.connect(self._on_load_finished)
            page.loadFinished.connect(self._on_load_finished_workaround)
        else:
            # for older Qt versions which break with the above
            page.loadProgress.connect(self._on_load_progress)
            page.loadFinished.connect(self._on_history_trigger)
            page.loadFinished.connect(self._restore_zoom)
            page.loadFinished.connect(self._on_load_finished)
```

**Comment by rcorre** [comment]
Never mind, I misunderstood the version check, and that's the bug you linked earlier anyways :)

**Comment by oceans112** [comment]
I'm having a hard time following you, but please let me know if I can further help you somehow. 
I cannot tell since this bug started to happen but it's there since quite a few version of qt and qutebrowser.

**Comment by rcorre** [comment]
@oceans112 no problem, I think we have all the details we need for now.

Seems like my last comment got lost: I'm seeing that we ignore all the `loadFinished` events (because `ok` is `True`), but we get two `loadProgress(100%)` events that each trigger a history insert.

**Comment by amacfie** [comment]
Is this still happening? Can it be deterministically reproduced?

**Comment by andrewcarlotti** [comment]
This was presumably fixed by #4699, and would remain fixed if the fix in #4699 is replaced by #5337.

**Comment by The-Compiler** [comment]
Assuming this is fixed then - thanks for the triage, @amacfie and @andrewcarlotti!
